#Задача предсказания. Линейная регрессия
**Задача регрессии** - предсказание числового признака

Предсказываемый признак называется **целевым**

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
#import frameworks
import pandas as pd
import seaborn as sns
from sklearn.linear_model import LinearRegression

In [ ]:
#Загрузка наших данных
df=pd.read_csv('/content/drive/MyDrive/Colab Notebooks/students.csv', delimiter=',')
#Создание укороченной таблицы т.к. мы ищем зависимость роста от размера ноги и наоборот
df_cut=df[['Growth', 'Shoe size']]
#Исключаем нулевые значения
df_cut=df_cut.dropna()

In [ ]:
sns.scatterplot(data=df_cut, x='Shoe size', y='Growth')

Гипотеза: рост линейно зависит от размера обуви 
Growth=a*ShSize+b

Коэффициенты a и b на данный момент не известны

Модель регрессии как раз умеется находить оптимальные коэффициенты

In [ ]:
LienearRegression = LinearRegression() #Создание "ИИ"
results=LienearRegression.fit(df_cut['Shoe size'].values.reshape(-1,1), y=df_cut['Growth'].values) 
#С помощью .fit() просим подогнать результаты
# Целевой признак - рост. Нецелевой признак - размер обуви.

In [ ]:
df_cut['Shoe size'].values

In [ ]:
df_cut['Shoe size'].values.reshape(-1,1)

In [ ]:
results.coef_, results.intercept_ #Вычисленные коэффициенты
#a=2.7 , b=60.37

(array([2.77717058]), 60.36581125736733)

Прогоним через нашу модель людей и для каждого из них она предскажет рост. И мы сравним предсказанный рост с фактическим ростом.

In [ ]:
df_cut['Predicted Growth']=results.predict(df_cut['Shoe size'].values.reshape(-1,1))

In [ ]:
from sklearn.metrics import mean_absolute_error #данная библиотека подключается для сравнения полученных значений и получения средней ошибки
mean_absolute_error(df_cut['Growth'], df_cut['Predicted Growth']) #Передаем истинный рост и предсказанный рост

3.450698681998386

In [ ]:
df_cut.head()

Growth  Shoe size  Predicted Growth
0     170         40        171.452635
1     191         43        179.784146
2     172         41        174.229805
3     168         38        165.898293
4     159         36        160.343952

#Посмотрим разницу в пальцах студентов

In [ ]:
df_finger=df[['MIddle and index finger','Middle and ring finger','Middle and little finger']]

In [ ]:
LienearRegression = LinearRegression() #Создание "ИИ"
results=LienearRegression.fit(df_finger[['Middle and ring finger','Middle and little finger']].values.reshape(-1,2), y=df_finger['MIddle and index finger'].values) 
results.intercept_, results.coef_

(7.436145518798954, array([0.10731496, 0.04389948]))

MIddle and index finger = 0.107*Middle and ring finger+0.04Middle and little finger+7.44

In [ ]:
df_finger['Predicted']=results.predict(df_finger[['Middle and ring finger','Middle and little finger']].values.reshape(-1,2))

<ipython-input-79-693785946f05>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_finger['Predicted']=results.predict(df_finger[['Middle and ring finger','Middle and little finger']].values.reshape(-1,2))


In [ ]:
mean_absolute_error(df_finger['Predicted'], df_finger['MIddle and index finger'])

3.094567105647603

#Проверка точности модели на новой выборке
До этого моменты мы тренировали модель и проверяли ее качество (mean_absolute_error) на одних и тех же данных.
Вообще говоря, это не есть корректно. Т.к. модель была построена с участием данных, на которых она потом проверяется.

Поэтому в анализе данных поступают так:

У вас есть две таблицы: тренировочная и тестовая выборка.
Модель строится по тренировочной выборке. Качество ее проверяется по тестовой выборке.
Сейчас мы это реализуем.

Загружаем вторую таблицу со студентами, которая не использовалась ранее:


In [ ]:
# загружаем данные тестовой выборки
df_test=pd.read_csv('/content/drive/MyDrive/Colab Notebooks/students_test.csv', delimiter=',')
df_test_finger=df_test[['MIddle and index finger','Middle and ring finger','Middle and little finger']]

ВНИМАНИЕ!

Через функцию predict прогоняем объекты **тестовой** выборки

In [ ]:
df_test_finger['Predicted']=results.predict(df_test_finger[['Middle and ring finger','Middle and little finger']].values.reshape(-1,2))

In [ ]:
mean_absolute_error(df_test_finger['Predicted'], df_test_finger['MIddle and index finger'])

4.680118432328658

#Задачки

Создайте таблицу df_cut, в которой будут только столбцы 'Age','Year of birth', 'Month of birthday' из файла students.csv . Запустите модель линейной регрессии, которая будет решать задачу предсказания признака 'Age' (целевой) по признакам 'Year of birth', 'Month of birthday' (нецелевые).

 

Когда модель будет натренирована, то загрузите тестовую выборку из файла students_test.csv 

 

Устройте предсказание возраста для студентов из тестовой выборки. 

 

В окошко ответа запишите возраст, предсказанный для первого по счету студента (то есть студента с индексом 0) из тестовой выборки.

In [ ]:
df_cut=df[['Age','Year of birth', 'Month of birthday']]

In [ ]:
LR = LinearRegression() #Создание "ИИ"
results=LienearRegression.fit(df_cut[['Year of birth', 'Month of birthday']].values.reshape(-1,2), y=df_cut['Age'].values) 
results.intercept_, results.coef_

(22.92993029669935, array([-0.00145442,  0.10005863]))

In [ ]:
# загружаем данные тестовой выборки
df_test=pd.read_csv('/content/drive/MyDrive/Colab Notebooks/students_test.csv', delimiter=',')
df_test_cut=df_test[['Age','Year of birth', 'Month of birthday']]

In [ ]:
df_test_cut['Predicted']=results.predict(df_test_cut[['Year of birth', 'Month of birthday']].values.reshape(-1,2))

<ipython-input-91-8d13248f9928>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_test_cut['Predicted']=results.predict(df_test_cut[['Year of birth', 'Month of birthday']].values.reshape(-1,2))


In [ ]:
df_test_cut.head()

В окошко ответа запишите среднюю абсолютную ошибку (mean absolute error) модели на тестовой выборке.

In [ ]:
mean_absolute_error(df_test_cut['Predicted'], df_test_cut['Age'])

1.6054960000660923